`RenderBox.{compute,get}DryBaseline`

LongCatIsLooong opened the pull request:
Introduce methods for computing the baseline location of a `RenderBox` without affecting the current layout (see [packages/flutter/lib/src/rendering/box.dart](https://github.com/flutter/flutter/pull/138369/files#diff-f2acffc9e582820293ac5a6dfbe0f8ac16491e992d4e89e074412b8d637f3cc6) for the signatures). 

In follow-up PRs, the new methods will be called mostly in `computeDryLayout` for RenderBoxes with baseline-aligned children. Unfortunately LayoutBuilders still don't have the ability to compute the intrinsic dimensions sideeffect-free.

Issues that depend on this: 
https://github.com/flutter/flutter/issues/138794
https://github.com/flutter/flutter/issues/71687

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

LongCatIsLooong committed:
`RenderBox.{compute,get}DryBaseline`

LongCatIsLooong committed:
tests

LongCatIsLooong committed:
Merge branch 'master' into text-intrinsics

LongCatIsLooong committed:
fix test failures

LongCatIsLooong committed:
Merge branch 'text-intrinsics' of github.com:LongCatIsLooong/flutter into text-intrinsics

LongCatIsLooong committed:
fix ListTile test failure

goderbauer reviewed (commented):
I only looked at `box.dart` so far.

goderbauer commented:
Can you remind me (and also add it to the PR description) what problem we can now solve by having the ability to compute baselines without non-destructively? Wasn't there some kind of an issue on file that we could link this PR to as a step towards resolving that issue?

LongCatIsLooong commented:
> Can you remind me (and also add it to the PR description) what problem we can now solve by having the ability to compute baselines without non-destructively? Wasn't there some kind of an issue on file that we could link this PR to as a step towards resolving that issue?

https://github.com/flutter/flutter/issues/71687 would be an example. But the PR by itself doesn't fix anything. The intrinsics methods have to be updated to use the new method(s) and I'm planning to do that in a different PR.

LongCatIsLooong commented on packages/flutter/lib/src/rendering/box.dart:
```diff
@@ -933,24 +933,123 @@ class BoxParentData extends ParentData {
+            ..['intrinsics dimension'] = dimension.name
+            ..['intrinsics argument'] = '$input',
+        _DryLayoutCache() => arguments..['getDryLayout constraints'] = '$input',
+        _DryBaselineCache() => arguments..['getDryBaseline constraints'] = '$input',
```
I kept the exact logic for fear of breaking the devtool. I'll try adding that.

goderbauer replied:
Isn't the baseline stuff all new? How can devtools already depend on it?

LongCatIsLooong commented on packages/flutter/lib/src/rendering/box.dart:
```diff
@@ -933,24 +933,123 @@ class BoxParentData extends ParentData {
   }
+}
+
+final class _DryBaselineCache implements _IntrinsicsCacheType<(BoxConstraints, TextBaseline), double?> {
```
I can try giving them better names and adding a short class comment explaining their roles in this.

LongCatIsLooong commented on packages/flutter/lib/src/rendering/box.dart:
```diff
@@ -2170,11 +2235,33 @@ abstract class RenderBox extends RenderObject {
-    _cachedBaselines ??= <TextBaseline, double?>{};
-    return _cachedBaselines!.putIfAbsent(baseline, () => computeDistanceToActualBaseline(baseline));
+    return _intrinsicsCache.getOrCompute(
+      _IntrinsicsCacheType.baseline,
```
> Why do we still offer the destructive one

There are some multi-child render box classes whose dry baseline implementation is relatively expensive (`RenderWrap` for example). The baseline values are cached, but they do require a full layout, and the layout results are discarded immediately after you get the baseline location (i.e. it's not reused by `performLayout`). So the destructive version is usually more performant as the children's offsets are already calculated and cached during `performLayout`.

>  interesting and hard to debug bugs where you get one layout if you don't call the dry methods and another if you do call them upfront.

You mean bugs in the render box subclass implementation? I had that problem while implementing the method for some subclasses. Had to refactor some of them to share the child positioning code so it's easy enough to see that the `performLayout` and the `computeDryBaseline` implementations are consistent. But I don't know of a good way to ensure this in an assert.

goderbauer replied:
> There are some multi-child render box classes whose dry baseline implementation is relatively expensive (RenderWrap for example). The baseline values are cached, but they do require a full layout, and the layout results are discarded immediately after you get the baseline location (i.e. it's not reused by performLayout). So the destructive version is usually more performant as the children's offsets are already calculated and cached during performLayout.

This should be discussed in the docs.

> You mean bugs in the render box subclass implementation? [...] But I don't know of a good way to ensure this in an assert.

Yes! Can we leverage the existing `debugCheckIntrinsicSizes` to do some spot checks that the implementations match?

Also, is there a way to have computeDistanceToActualBaseline and computeDryBaseline share more code so you don't have to implement the same logic twice?

LongCatIsLooong commented on packages/flutter/lib/src/rendering/box.dart:
```diff
@@ -2170,11 +2235,33 @@ abstract class RenderBox extends RenderObject {
+  /// non-debug modes.
+  @nonVirtual
+  @visibleForTesting
+  double? debugGetDistanceToBaseline(TextBaseline baseline) {
```
`getDistanceToBaseline` can't be called in tests (there are a few asserts that would fire if you call that method during the idle phase).

goderbauer replied:
Add that information to the doc comment?

LongCatIsLooong commented on packages/flutter/lib/src/rendering/box.dart:
```diff
@@ -2353,42 +2440,20 @@ abstract class RenderBox extends RenderObject {
 
-  @override
-  void layout(Constraints constraints, {bool parentUsesSize = false}) {
-    if (hasSize && constraints != this.constraints &&
-        _cachedBaselines != null && _cachedBaselines!.isNotEmpty) {
-      // The cached baselines data may need update if the constraints change.
-      _cachedBaselines?.clear();
-    }
```
Yeah exactly. The cache now only has to be invalidated when the intrinsic state of the render box changes.

goderbauer commented:
> https://github.com/flutter/flutter/issues/71687 would be an example. But the PR by itself doesn't fix anything. The intrinsics methods have to be updated to use the new method(s) and I'm planning to do that in a different PR.

Could you still mention that issue in the PR description and indicate that this is work towards fixing the issue to leave some breadcrumbs/context for future archeologists?

goderbauer commented on packages/flutter/test/rendering/baseline_utils.dart:
```diff
@@ -0,0 +1,76 @@
+      }),
+      paintsNothing,
+    ),
+  );
```
Help me, what is this expect supposed to be asserting?

LongCatIsLooong replied:
Sometimes the render object doesn't paint anything so that has to be matched using `paintsNothing`.

goderbauer commented on packages/flutter/test/rendering/baseline_utils.dart:
```diff
@@ -0,0 +1,76 @@
+  }
+}
+
+double? verifyDryBaseline(RenderBox box) {
```
Just from the method name I would have expected that this method would only contain the expect on line 57. It is not clear to me what the beforePaintInstructions/afterPaintInstructions and their associated expects are supposed to test? And the return value is also somewhat surprising (to me it implies that I need to compare it with something in my test in an expect, but the expects seem to be build into this methods...)

LongCatIsLooong replied:
I'm considering getting rid of this helper and banking on that the assert in `debugPaint` would catch any inconsistencies here.

goderbauer commented on packages/flutter/test/rendering/baseline_utils.dart:
```diff
@@ -0,0 +1,76 @@
+import 'package:flutter/rendering.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+class _PaintInstructionMatcher extends Matcher {
```
Does this matcher provide a clear error message when it fails? I see a lot of our custom matcher implement describeMismatch to help out with that, but this one doesn't.

goderbauer reviewed (commented):
Review of the rest.

goderbauer commented:
How did you decide which RenderBoxes need a custom computeDryBaseline implementation? How should I as a custom render object author determine that I need one (the latter should be added to the docs somewhere)?

LongCatIsLooong committed:
Merge remote-tracking branch 'upstream/main' into text-intrinsics

LongCatIsLooong committed:
Merge remote-tracking branch 'upstream/main' into text-intrinsics

LongCatIsLooong committed:
nav bar

LongCatIsLooong committed:
Merge branch 'text-intrinsics' of github.com:LongCatIsLooong/flutter into text-intrinsics

LongCatIsLooong committed:
wip

LongCatIsLooong committed:
wip

LongCatIsLooong committed:
wip

LongCatIsLooong committed:
wip

LongCatIsLooong committed:
Merge remote-tracking branch 'upstream/main' into text-intrinsics

LongCatIsLooong committed:
fix RenderWrap

LongCatIsLooong committed:
fix more implementations

LongCatIsLooong committed:
fix more implementations

LongCatIsLooong committed:
Merge remote-tracking branch 'upstream/main' into text-intrinsics

LongCatIsLooong commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -688,36 +682,18 @@ class RenderParagraph extends RenderBox with ContainerRenderObjectMixin<RenderBo
+    // (paint, getOffsetForCaret, to name a few) must not assume the TextPainter
+    // is laid out with the correct constraints. These methods should typically
+    // call _layoutTextWithConstraints before querying the TextPainter, to make
+    // sure the TextPainter is laid out using the right constraints.
```
I don't think there's something that remotely resembles a "system", we just call `_layoutText` excessively every time the text layout gets accessed.

LongCatIsLooong committed:
review

LongCatIsLooong committed:
fix RenderAnimatedSize

LongCatIsLooong committed:
fix RenderAspectRatio

LongCatIsLooong committed:
review

LongCatIsLooong committed:
Merge remote-tracking branch 'upstream/main' into text-intrinsics

LongCatIsLooong committed:
fix dem tests

LongCatIsLooong committed:
revert test changes

LongCatIsLooong committed:
syntax

LongCatIsLooong committed:
revert more tests

LongCatIsLooong committed:
rename

LongCatIsLooong committed:
Merge remote-tracking branch 'upstream/main' into text-intrinsics

LongCatIsLooong commented:
I think I've addressed most comments. Right now in debug mode, the `RenderBox.getDryBaseline` method still returns a valid baseline offset even if there's a LayoutBuilder descendent, if the `RenderBox` has been previously laid out at that input BoxConstraints and the baseline offset cache hasn't been invalidated. To fix that we'll need to distinguish "dry" value from "wet" value in the cache (but only in debug mode) so I'm not sure if it's worth implementing.

LongCatIsLooong committed:
Merge remote-tracking branch 'upstream/main' into text-intrinsics

LongCatIsLooong commented:
Splitting this into smaller PRs.
